yarpl/include/yarpl/flowable/FlowableOperator.h

```diff
@@ -55,48 +55,110 @@ class FlowableOperator : public Flowable<D> {
     }
 
     void subscriberOnNext(D value) {
-      subscriber_->onNext(std::move(value));
-    }
-
-    void request(int64_t delta) override {
-      upstream_->request(delta);
+      if (subscriber_) {
+        subscriber_->onNext(std::move(value));
+      }
     }
 
-    // this method should be used to terminate the operators
+    /// Terminates both ends of an operator normally.
     void terminate() {
-      subscriber_->onComplete(); // should break the cycle to this
-      upstream_->cancel(); // should break the cycle to this
+      if (isTerminated()) {
+        return;
+      }
+
+      if (auto upstream = std::move(upstream_)) {
+        upstream->cancel();
+      }
+      if (auto subscriber = std::move(subscriber_)) {
+        subscriber->onComplete();
+      }
+
       Refcounted::decRef(*this);
     }
 
-    void cancel() override {
-      assert(upstream_ && "subscription was already terminated");
-      (decltype(upstream_)(std::move(upstream_)))->cancel();
-      subscriber_.reset(); // breaking the cycle
+    /// Terminates both ends of an operator with an error.
+    void terminateErr(std::exception_ptr eptr) {
+      if (isTerminated()) {
+        return;
+      }
+
+      if (auto upstream = std::move(upstream_)) {
+        upstream->cancel();
+      }
+      if (auto subscriber = std::move(subscriber_)) {
+        subscriber->onError(std::move(eptr));
+      }
+
       Refcounted::decRef(*this);
     }
 
-    void onComplete() override {
-      assert(upstream_ && "subscription was already terminated");
-      upstream_.reset(); // breaking the cycle
-      subscriber_->onComplete();
-      Refcounted::decRef(*this);
+    // Subscription.
+
+    void request(int64_t delta) override {
+      if (upstream_) {
```

Comment:
I don't believe so, because if `upstream_ == nullptr` then we've terminated this operator.  Any requests still flowing in from downstream are "late" so to speak, and should be dropped safely.